### PR TITLE
Enable scan build

### PR DESCRIPTION
**Pull request by coryan**
This time it might work.  This is related to #10.  Enables a Clang build, it takes "only" ~30min, it uses a installed gRPC to speed up things. Caching has not worked.

### CODE REVIEW

File: ci/Dockerfile.ubuntu-with-grpc [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+
+ARG DISTRO=ubuntu
+ARG DISTRO_VERSION=16.04
+FROM cached-${DISTRO}-${DISTRO_VERSION}:tip AS staging
+MAINTAINER "Carlos O'Ryan <coryan@google.com>"
+
+RUN apt-get install -y automake libtool build-essential
+
+ARG NCPU=2
+ARG CXX=g++
```
Comment (mbrukman) [inline ci/Dockerfile.ubuntu-with-grpc]: If you care about speed of compilation, I believe Clang should be faster in compiling than GCC, if that's on your critical path.

File: .travis.yml [hunk 2]
```diff
@@ -47,15 +48,22 @@ matrix:
     - os: linux
       compiler: gcc
       env: DISTRO=centos DISTRO_VERSION=7 BUILD_TYPE=Release
+    - # This is a fairly specialized entry in the matrix.  It builds the clients
```
Comment (mbrukman) [inline .travis.yml]: Do you want to only run this on the `master` branch, and not any of the PRs, to speed up the review / merge process? Or is the goal to run this on every PR (and every PR revision)?

Comment (coryan) [reply .travis.yml]: Good point.  Let's skip it for PRs.  Too slow, like the macOS build.

File: .travis.yml [hunk 3]
```diff
@@ -47,15 +48,22 @@ matrix:
...
+      # using a pre-installed gRPC package.  To do so, it builds the normal
+      # docker image with all the build tools, and uses that image to build and
```
[the remark this answers is not in the document]
Comment (coryan) [reply .travis.yml]: Fixed.

File: .travis.yml [hunk 4]
```diff
@@ -47,15 +48,22 @@ matrix:
...
+      # install gRPC.
+      os: linux
+      compiler: clang
+      env: DISTRO=ubuntu-with-grpc DISTRO_VERSION=16.04 BUILD_TYPE=Release
```
[the remark this answers is not in the document]
Comment (coryan) [reply .travis.yml]: Black magic indeede.  That was a fortunate mistake. I thought it was needed, but turns out it is not:

https://travis-ci.org/coryan/google-cloud-cpp/jobs/322665795#L517

I added the `\` for consistency, we should figure out how we want these lines to look.

File: CMakeLists.txt [hunk 5]
```diff
@@ -100,8 +100,6 @@ elseif ("${GOOGLE_CLOUD_CPP_GRPC_PROVIDER}" STREQUAL "package")
         ${PROTOBUF_SRC_ROOT_FOLDER}/vsprojects/${_PROTOBUF_ARCH_DIR}Debug
         )
     mark_as_advanced(PROTOC_GRPCPP_PLUGIN_EXECUTABLE)
-    # googletest
-    add_subdirectory(googletest EXCLUDE_FROM_ALL)
```
[the remark this answers is not in the document]
Comment (coryan) [reply CMakeLists.txt]: That branch of the code was never tested.  The path to `googletest` was wrong should have been `third_party/googletest`.  I think we should cleanup this when we work on #127 and #126.

File: ci/Dockerfile.ubuntu-with-grpc [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+ARG DISTRO=ubuntu
+ARG DISTRO_VERSION=16.04
+FROM cached-${DISTRO}-${DISTRO_VERSION}:tip AS staging
+MAINTAINER "Carlos O'Ryan <coryan@google.com>"
+
+RUN apt-get install -y automake libtool build-essential
```
[the remark this answers is not in the document]
Comment (coryan) [reply ci/Dockerfile.ubuntu-with-grpc]: Better yet, move that to `Dockerfile.ubuntu`.

File: ci/Dockerfile.ubuntu-with-grpc (hunk 1 above)
Comment (coryan) [reply ci/Dockerfile.ubuntu-with-grpc]: Done.
Thought these are the defaults and the `.travis.yml` file changed them to `clang++` and `clang`.

File: cmake/IncludeAbseil.cmake [hunk 7]
```diff
@@ -15,27 +15,18 @@
 include(${PROJECT_SOURCE_DIR}/cmake/IncludeCctz.cmake)
 include(${PROJECT_SOURCE_DIR}/cmake/IncludeGMock.cmake)
 
-set(GOOGLE_CLOUD_CPP_ABSEIL_PROVIDER "module" CACHE STRING "How to find the abseil libraries")
-set_property(CACHE GOOGLE_CLOUD_CPP_ABSEIL_PROVIDER PROPERTY STRINGS "module" "package")
+# Depending on how gRPC is used (module vs. package) the gtest target may be
```
[the remark this answers is not in the document]
Comment (coryan) [reply cmake/IncludeAbseil.cmake]: Fixed.

File: cmake/IncludeAbseil.cmake [hunk 8]
```diff
@@ -15,27 +15,18 @@
...
+# already defined, if it is, we cannot redefine it.
```
[the remark this answers is not in the document]
Comment (coryan) [reply cmake/IncludeAbseil.cmake]: Fixed.

File: cmake/IncludeCctz.cmake [hunk 9]
```diff
@@ -12,31 +12,24 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-set(GOOGLE_CLOUD_CPP_CCTZ_PROVIDER "module" CACHE STRING "How to find the cctz libraries")
-set_property(CACHE GOOGLE_CLOUD_CPP_CCTZ_PROVIDER PROPERTY STRINGS "module" "package")
+# While the documentation is not clear as to whether cctz can be used as an
+# installed library, we only use it as part of abseil, which is supposed to
```
[the remark this answers is not in the document]
Comment (coryan) [reply cmake/IncludeCctz.cmake]: Fixed.

File: cmake/IncludeGMock.cmake [hunk 10]
```diff
@@ -12,6 +12,9 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
+# abseil requires a gmock target to be defined.  Unlike the gtest target gRPC
```
[the remark this answers is not in the document]
Comment (coryan) [reply cmake/IncludeGMock.cmake]: Fixed.

**Comment by coryan** [comment]
PTAL.

**Pull request merged**